### PR TITLE
Add support for Fourmolu 0.8

### PR DESCRIPTION
**Pull request by brandonchinn178**
Also includes some clean up. Will we always be supporting every version of Fourmolu, or will HLS ever remove support for old versions of Fourmolu?

cc @parsonsmatt @georgefst 

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3078"></a>

### CODE REVIEW

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu/Shim.hs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    , cfgFileFixities = mempty
+    }
+#endif
+
+emptyConfig :: FourmoluConfig
+emptyConfig =
+  FourmoluConfig
+    { cfgFilePrinterOpts = mempty
+    , cfgFileFixities = mempty
+    }
```
Comment (brandonchinn178) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu/Shim.hs:46]: will eventually be under a `#if !MIN_VERSION_fourmolu(0,8,1)` macro: https://github.com/fourmolu/fourmolu/pull/221

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs [hunk 2]
```diff
@@ -100,8 +100,8 @@ provider recorder plId ideState typ contents fp fo = withIndefiniteProgress titl
                         pure . Left . responseError $ "Fourmolu failed with exit code " <> T.pack (show n)
         else do
             let format fourmoluConfig =
-                    first (mkError . show)
-                        <$> try @OrmoluException (makeDiffTextEdit contents <$> ormolu config fp' (T.unpack contents))
+                    bimap (mkError . show) (makeDiffTextEdit contents)
+                        <$> try @OrmoluException (ormolu config fp' (T.unpack contents))
```
Comment (georgefst) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:103]: Nice. Surpised I didn't notice that opportunity.

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu/Shim.hs [hunk 3]
```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE CPP #-}
+
+module Ide.Plugin.Fourmolu.Shim (
```
Comment (georgefst) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu/Shim.hs:3]: Excellent. This is much cleaner.

**Review by georgefst — APPROVED** [review]
Thanks!

**Comment by georgefst** [comment]
> Will we always be supporting every version of Fourmolu, or will HLS ever remove support for old versions of Fourmolu?

I don't see a reason to keep _really_ old versions around, especially if it means huge amounts of CPP. But it's likely that some people won't want to be forced in to updating Fourmolu (and thus probably reformatting their codebase) by an HLS update\*. Then again, part of the reason for adding the CLI mode was to give users that flexibility.

I've wanted to remove some backwards compatibility before, but it wasn't possible then because of constraints from HLS' other dependencies: https://github.com/haskell/haskell-language-server/pull/2951.

\* Although in practice, most people are getting HLS from bindists, so they don't have much choice without building from source instead.

**Comment by brandonchinn178** [comment]
@georgefst By "building from source", do you mean fourmolu? Now that we release binaries on the releases page, is there some hls logic that can download fourmolu binaries where possible instead of building fourmolu? (Even better, make that the canonical way; dont bundle fourmolu in hls, just have hls provide a dropdown to install version of fourmolu)

Keeping old versions isnt terrible right now, because all the cpp is for the 0.7 upgrade. But I think maybe once 0.7 becomes two years old (or something) itd be nice to drop support

**Comment by georgefst** [comment]
> @georgefst By "building from source", do you mean fourmolu?

I meant HLS, but I suppose that statement works for either, once you take the CLI option in to account.

We _could_ make downloading the binary the default, but I'm more confident of correctness when using the library. And it would go against the way every other HLS plugin is doing things (although that is not, in itself, a great argument against).

**Pull request merged**